File: labs/ch01-python-basics/src/02_functions_and_types.py
"""02 · 函数与类型提示 — 把 Java 方法签名翻译成 Python.

对照表:
    Java                            Python
    ──────                          ──────
    public String foo(int x)         def foo(x: int) -> str
    public List<String> bar()        def bar() -> list[str]
    public Optional<User> find(id)   def find(id: str) -> User | None
    public void baz(int... args)     def baz(*args: int) -> None
    public void qux(Map<K,V> kv)     def qux(**kv: int) -> None
    method overloading 重载           不支持, 用默认参数 / @overload 模拟

Run:
    python 02_functions_and_types.py
"""

from __future__ import annotations


# ===== 1. 基础函数 =====
def add(x: int, y: int) -> int:
    """两数相加 — type hints 可选, 但强烈推荐写."""
    return x + y


# ===== 2. 默认参数 (Java 没有, 必须重载方法) =====
def greet(name: str, greeting: str = "Hello") -> str:
    """默认参数让"重载"变成一行代码."""
    return f"{greeting}, {name}!"


# ===== 3. 可变参数 (Java 的 String...) =====
def sum_all(*numbers: int) -> int:
    """*numbers 收集任意个位置参数为 tuple."""
    return sum(numbers)


# ===== 4. 关键字参数 (Java 没有直接对应, 类似 Map<String, Object>) =====
def build_user(**attrs: str) -> dict[str, str]:
    """**attrs 收集任意个 key=value 为 dict."""
    return attrs


# ===== 5. 多返回值 — Python 用 tuple, Java 要自己 wrap 一个类 =====
def divmod_pair(a: int, b: int) -> tuple[int, int]:
    """同时返回商和余数 — 调用方可解构."""
    return a // b, a % b


# ===== 6. 可空类型 — Python 3.10+ 用 X | None =====
def find_user(user_id: str) -> dict[str, str] | None:
    """模拟一个可能找不到用户的查询 — 对照 Java 的 Optional<User>."""
    if user_id == "alice":
        return {"id": "alice", "role": "admin"}
    return None


# ===== 7. 高阶函数 — 函数当参数, Java 要 Functional Interface =====
def apply(func, value):
    """对照 Java 8 lambda: Function<T, R>.apply(t)."""
    return func(value)


# ===== 8. lambda — 一行匿名函数 =====
square = lambda x: x * x  # noqa: E731


# ===== 9. 闭包 — 函数捕获外部变量 =====
def make_counter(start: int = 0):
    """闭包: 内层函数捕获外层 count, 每次调用 += 1."""
    count = start

    def inc() -> int:
        nonlocal count  # 修改外层变量必须用 nonlocal
        count += 1
        return count

    return inc


def main() -> None:
    print(f"add(2, 3)             = {add(2, 3)}")
    print(f"greet('Alice')        = {greet('Alice')}")
    print(f"greet('Bob', '你好')  = {greet('Bob', '你好')}")
    print(f"sum_all(1,2,3,4,5)    = {sum_all(1, 2, 3, 4, 5)}")
    print(f"build_user(...)       = {build_user(name='alice', role='admin')}")
    print(f"divmod_pair(17, 5)    = {divmod_pair(17, 5)}")
    print(f"find_user('alice')    = {find_user('alice')}")
    print(f"find_user('zzz')      = {find_user('zzz')}")
    print(f"apply(square, 7)      = {apply(square, 7)}")

    counter = make_counter(10)
    print(f"counter() x3          = {counter()}, {counter()}, {counter()}")


if __name__ == "__main__":
    main()

File: labs/ch01-python-basics/src-java/02_FunctionsAndTypes.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.IntSupplier;
import java.util.stream.IntStream;

/**
 * 等价 Python: src/02_functions_and_types.py — 看这个文件感受 JDK 23 在做和 Python 同样的事.
 *
 * <p>核心对照:
 * <pre>
 *   Java 8 老写法                       JDK 23 新写法                          Python 写法
 *   ─────────────                       ──────────────                        ────────────
 *   public String foo(int x)            String foo(int x)                     def foo(x: int) -> str
 *   List&lt;String&gt; bar()                  返回 record / 不变                     def bar() -> list[str]
 *   Optional&lt;User&gt; find(id)             Optional&lt;User&gt; (or sealed result)     def find(id) -> User | None
 *   void baz(int... args)               void baz(int... args)                 def baz(*args: int)
 *   Map&lt;String, Object&gt; kv               record 或 Map.of                      **kwargs
 *   方法重载 (overload)                   仍是 overload (Python 不支持)           默认参数模拟
 *   Function&lt;T,R&gt; (Java 8 lambda)        同左 + Stream API                      def / lambda
 * </pre>
 *
 * <p>JDK 23 特性使用:
 * <ul>
 *   <li>{@code record} — 多返回值 (Python tuple 等价)</li>
 *   <li>{@code var} — 局部类型推断</li>
 *   <li>Stream API — 替代命令式 for 循环</li>
 *   <li>Pattern matching {@code instanceof}</li>
 * </ul>
 */
class FunctionsAndTypes {

    // ===== 1. 基础函数 =====
    static int add(final int x, final int y) {
        return x + y;
    }

    // ===== 2. 默认参数 — Java 没有, 必须 overload (Python 一行搞定) =====
    static String greet(final String name) {
        return greet(name, "Hello");
    }

    static String greet(final String name, final String greeting) {
        return "%s, %s!".formatted(greeting, name);
    }

    // ===== 3. 可变参数 — Python 的 *args =====
    static int sumAll(final int... numbers) {
        return IntStream.of(numbers).sum();
    }

    // ===== 4. 关键字参数 — Java 没有原生 **kwargs, 用 Map 模拟 =====
    static Map<String, String> buildUser(final Map<String, String> attrs) {
        return new HashMap<>(attrs);
    }

    // ===== 5. 多返回值 — JDK 23 用 record 替代 Python tuple =====
    /** 商和余数的多元返回 — 等价 Python 的 tuple[int, int]. */
    record DivMod(int quotient, int remainder) {}

    static DivMod divmodPair(final int a, final int b) {
        return new DivMod(a / b, a % b);
    }

    // ===== 6. 可空类型 — Java 用 Optional, Python 用 X | None =====
    /** 用户记录 — 等价 Python 的 dict[str, str], 但有类型. */
    record User(String id, String role) {}

    static Optional<User> findUser(final String userId) {
        return "alice".equals(userId)
                ? Optional.of(new User("alice", "admin"))
                : Optional.empty();
    }

    // ===== 7. 高阶函数 — Java 8 Function<T,R>, Python 函数本身就是一等公民 =====
    static <T, R> R apply(final Function<T, R> func, final T value) {
        return func.apply(value);
    }

    // ===== 8. lambda — Java 8 起就有, JDK 23 没变, 但 Stream 用起来更顺 =====
    static final Function<Integer, Integer> SQUARE = x -> x * x;

    // ===== 9. 闭包 — Java lambda 只能捕获 effectively final, 想要可变状态需 wrap =====
    static IntSupplier makeCounter(final int start) {
        // 用单元素数组绕过 effectively final 限制 (Python 用 nonlocal)
        final int[] count = {start};
        return () -> ++count[0];
    }

    public static void main(final String[] args) {
        System.out.println("add(2, 3)             = " + add(2, 3));
        System.out.println("greet('Alice')        = " + greet("Alice"));
        System.out.println("greet('Bob', '你好')  = " + greet("Bob", "你好"));
        System.out.println("sumAll(1..5)          = " + sumAll(1, 2, 3, 4, 5));

        // Map.of 是 Java 9+ 的不可变 Map 字面量, 类似 Python dict literal
        final var user = buildUser(Map.of("name", "alice", "role", "admin"));
        System.out.println("buildUser(...)        = " + user);

        // record 析构 — JDK 21+ 支持 record pattern
        final DivMod dm = divmodPair(17, 5);
        System.out.println("divmodPair(17, 5)     = " + dm);

        // pattern matching switch — JDK 21+ stable
        final var found = findUser("alice");
        System.out.println("findUser('alice')     = " + describeOptional(found));
        System.out.println("findUser('zzz')       = " + describeOptional(findUser("zzz")));

        System.out.println("apply(SQUARE, 7)      = " + apply(SQUARE, 7));

        final var counter = makeCounter(10);
        System.out.println("counter() x3          = "
                + counter.getAsInt() + ", " + counter.getAsInt() + ", " + counter.getAsInt());

        // Stream API — Python 列表推导式的 Java 等价
        final List<Integer> squared = List.of(1, 2, 3, 4, 5).stream()
                .map(SQUARE)
                .toList();
        System.out.println("squared (Stream)      = " + squared);
    }

    /** Pattern matching switch (JDK 21+ stable) — 比 Optional.map 更直观. */
    static String describeOptional(final Optional<User> maybeUser) {
        return switch (maybeUser.orElse(null)) {
            case null -> "<not found>";
            case User u -> u.toString();
        };
    }
}
